Applications/Spire/Include/Spire/Spire/AnyRef.hpp

```diff
@@ -0,0 +1,355 @@
+
+      /** Constructs an AnyRef referencing a non-const, non-volatile object. */
+      template<typename T>
+      requires std::conjunction_v<
+        std::negation<std::is_same<std::decay_t<T>, AnyRef>>,
+        std::is_copy_constructible<std::decay_t<T>>>
+      AnyRef(T& ref) noexcept;
+
+      /** Constructs an AnyRef referencing a const object. */
+      template<typename T>
+      requires std::conjunction_v<
+        std::negation<std::is_same<std::decay_t<T>, AnyRef>>,
+        std::is_copy_constructible<std::decay_t<T>>>
+      AnyRef(const T& ref) noexcept;
+
+      /** Constructs an AnyRef referencing a volatile object. */
+      template<typename T>
+      requires std::conjunction_v<
+        std::negation<std::is_same<std::decay_t<T>, AnyRef>>,
+        std::is_copy_constructible<std::decay_t<T>>>
+      AnyRef(volatile T& ref) noexcept;
+
+      /** Constructs an AnyRef referencing a const and volatile object. */
+      template<typename T>
+      requires std::conjunction_v<
+        std::negation<std::is_same<std::decay_t<T>, AnyRef>>,
+        std::is_copy_constructible<std::decay_t<T>>>
+      AnyRef(const volatile T& ref) noexcept;
+
+      AnyRef(const AnyRef& any) noexcept = default;
+
+      AnyRef(AnyRef&& any) noexcept;
+
+      /** Returns <code>true</code> iff this object is not empty. */
+      bool has_value() const noexcept;
+
+      /** Returns the typeid of the referenced object. */
+      const std::type_info& get_type() const noexcept;
+
+      /** Returns <code>true</code> iff the referenced object is const. */
+      bool is_const() const noexcept;
+
+      /** Returns <code>true</code> iff the referenced object is volatile. */
+      bool is_volatile() const noexcept;
+
+      /**
+       * Returns <code>true</code> iff the referenced object is both const and
+       * volatile.
+       */
+      bool is_const_volatile() const noexcept;
+
+      AnyRef& operator =(const AnyRef& any) noexcept = default;
+
+      AnyRef& operator =(std::nullptr_t) noexcept;
+
+      template<typename T>
+      AnyRef& operator =(T& ref) noexcept;
+
+      template<typename T>
+      AnyRef& operator =(const T& ref) noexcept;
+
+      template<typename T>
+      AnyRef& operator =(volatile T& ref) noexcept;
+
+      template<typename T>
+      AnyRef& operator =(const volatile T& ref) noexcept;
+
+      AnyRef& operator =(AnyRef&& any) noexcept;
+
+    private:
+      struct AnyRefBase {
+        virtual void* get() = 0;
+        virtual const void* get_const() const = 0;
+        virtual volatile void* get_volatile() const = 0;
+        virtual const volatile void* get_const_volatile() const = 0;
+        virtual const std::type_info& get_type() const = 0;
+        virtual constexpr bool is_const() const = 0;
+        virtual constexpr bool is_volatile() const = 0;
+        virtual constexpr bool is_const_volatile() const = 0;
+        virtual ~AnyRefBase() = default;
+      };
+      template<typename T>
```

Comment:
Look at implementing this without these classes, this class is intended to be very lightweight and efficient.